perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,717 @@
+      size *= dim.d[j];
+    }
+    size *= dataTypeToSize(trt_engine_->getBindingDataType(i));
+    trt_buffer_sizes_[i] = size;
+    CHECK_CUDA(cudaMalloc(&trt_buffer_dev_[i], size));
+  }
+
+  RCLCPP_INFO(rclcpp::get_logger("BEVDet"), "img num binding : %d", trt_engine_->getNbBindings());
+
+  post_buffer_ = reinterpret_cast<void **>(new void *[class_num_pre_task_.size() * 6]);
+  for (size_t i = 0; i < class_num_pre_task_.size(); i++) {
+    post_buffer_[i * 6 + 0] = trt_buffer_dev_[buffer_map_["reg_" + std::to_string(i)]];
+    post_buffer_[i * 6 + 1] = trt_buffer_dev_[buffer_map_["height_" + std::to_string(i)]];
+    post_buffer_[i * 6 + 2] = trt_buffer_dev_[buffer_map_["dim_" + std::to_string(i)]];
+    post_buffer_[i * 6 + 3] = trt_buffer_dev_[buffer_map_["rot_" + std::to_string(i)]];
+    post_buffer_[i * 6 + 4] = trt_buffer_dev_[buffer_map_["vel_" + std::to_string(i)]];
+    post_buffer_[i * 6 + 5] = trt_buffer_dev_[buffer_map_["heatmap_" + std::to_string(i)]];
+  }
+
+  return;
+}
+
+void BEVDet::initViewTransformer(
+  std::shared_ptr<int> & ranks_bev_ptr, std::shared_ptr<int> & ranks_depth_ptr,
+  std::shared_ptr<int> & ranks_feat_ptr, std::shared_ptr<int> & interval_starts_ptr,
+  std::shared_ptr<int> & interval_lengths_ptr)
+{
+  int num_points = n_img_ * depth_num_ * feat_h_ * feat_w_;
+  Eigen::Vector3f * frustum = new Eigen::Vector3f[num_points];
+
+  for (int i = 0; i < n_img_; i++) {
+    for (int d_ = 0; d_ < depth_num_; d_++) {
+      for (int h_ = 0; h_ < feat_h_; h_++) {
+        for (int w_ = 0; w_ < feat_w_; w_++) {
+          int offset =
+            i * depth_num_ * feat_h_ * feat_w_ + d_ * feat_h_ * feat_w_ + h_ * feat_w_ + w_;
+          (frustum + offset)->x() = static_cast<float>(w_) * (input_img_w_ - 1) / (feat_w_ - 1);
+          (frustum + offset)->y() = static_cast<float>(h_) * (input_img_h_ - 1) / (feat_h_ - 1);
+          (frustum + offset)->z() = static_cast<float>(d_) * depth_step_ + depth_start_;
+
+          // eliminate post transformation
+          *(frustum + offset) -= post_trans_.translation();
+          *(frustum + offset) = post_rot_.inverse() * *(frustum + offset);
+          //
+          (frustum + offset)->x() *= (frustum + offset)->z();
+          (frustum + offset)->y() *= (frustum + offset)->z();
+          // img to ego -> rot -> trans
+          *(frustum + offset) = cams2ego_rot_[i] * cams_intrin_[i].inverse() * *(frustum + offset) +
+                                cams2ego_trans_[i].translation();
+
+          // voxelization
+          *(frustum + offset) -= Eigen::Vector3f(x_start_, y_start_, z_start_);
+          (frustum + offset)->x() = static_cast<int>((frustum + offset)->x() / x_step_);
+          (frustum + offset)->y() = static_cast<int>((frustum + offset)->y() / y_step_);
+          (frustum + offset)->z() = static_cast<int>((frustum + offset)->z() / z_step_);
+        }
+      }
+    }
+  }
+
+  int * _ranks_depth = new int[num_points];
```

Comment:
Allocate `_ranks_depth` and `_ranks_feat` using `std::unique_ptr<int[]>` for improved memory management.